protocols/observed/src/lib.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Connection upgrade to allow retrieving the externally visible address (as dialer) or
+//! to report the externally visible address (as listener).
+
+extern crate bytes;
+extern crate futures;
+extern crate libp2p_core;
+extern crate tokio_codec;
+extern crate tokio_io;
+extern crate unsigned_varint;
+
+use bytes::Bytes;
+use futures::{future, prelude::*};
+use libp2p_core::{ConnectionUpgrade, Endpoint, Multiaddr};
+use std::{io, iter};
+use tokio_codec::{FramedRead, FramedWrite};
+use tokio_io::{AsyncRead, AsyncWrite};
+use unsigned_varint::codec::UviBytes;
+
+/// The output, this connection upgrade produces.
+pub enum Output<C> {
+    /// As `Dialer`, we get our own externally observed address.
+    Address(Multiaddr),
+    /// As `Listener`, we return a sender which allows reporting the observed
+    /// address the client.
+    Sender(Sender<C>)
+}
+
+/// `Sender` allows reporting back the observed address to the remote endpoint.
+pub struct Sender<C> {
+    io: FramedWrite<C, UviBytes>
+}
+
+impl<C: AsyncWrite> Sender<C> {
+    /// Send address `a` to remote as the observed address.
+    pub fn send_address(self, a: &Multiaddr) -> impl Future<Item=C, Error=io::Error> {
```

Comment:
I don't think this should have a return value. We want to enforce the protocol's correct behaviour (which is to close the substream afterwards) in the API.

Also, taking ownership of the `Multiaddr` removes one copy.